resdk/tests/functional/data_upload/e2e_upload.py

```diff
@@ -0,0 +1,268 @@
+        self.assertEqual(sample_2b.descriptor, ann_2)
+        self.assertEqual(sample_3.descriptor, ann_3)
+        self.assertEqual(sample_4.descriptor, {})
+        self.assertEqual(sample_5.descriptor, {})
+
+    def test_export(self):
+        resdk_logger.start_logging()
+
+        # Create the collection with named, unannotated samples
+        collection = self.res.collection.create(name='Test export annotation')
+
+        reads_1, reads_2 = self.get_reads(2, collection)
+
+        # Two different samples
+        sample_1 = reads_1.sample
+        sample_1.name = 'Sample 1'
+        sample_1.save()
+
+        sample_2 = reads_2.sample
+        sample_2.name = 'Sample 2'
+        ann_2 = {
+            'sample': {
+                'genotype': '',
+                'cell_type':
+                'Mixed',
+                'optional_char': [
+                    'LIBRARY_STRATEGY:Illumina Standard Prep ',
+                    'OTHER_CHAR_1:2x75 bp',
+                    'OTHER_CHAR_2:subdural cortical tumor, frontal/lateral'
+                    ' location. Easily isolated sample',
+                    'TISSUE:Tumor'
+                ],
+                'strain': 'N/A',
+                'source': 'Tumor',
+                'organism': 'Homo sapiens',
+                'molecule': 'total RNA',
+                'annotator': 'Tristan Brown',
+                'description': '',
+            }
+        }
+        sample_2.descriptor_schema = 'sample'
+        sample_2.descriptor = ann_2
+        sample_2.save()
+
+        reads_ann = {
+            'experiment_type': 'RNA-Seq',
+            'protocols': {
+                'growth_protocol': 'N/A',
+                'treatment_protocol': 'Control'
+            }
+        }
+        reads_2.descriptor_schema = 'reads'
+        reads_2.descriptor = reads_ann
+        reads_2.save()
+
+        # Export the new template
+        filepath = 'template1.xlsm'
+        try:
+            os.remove(filepath)
+        except OSError:
+            pass
+
+        collection.export_annotation(filepath)
+        assert os.path.exists(filepath)
+        # TODO: Include a file hash check
+        os.remove(filepath)
+
+    def test_upload_reads(self):
+        resdk_logger.start_logging()
+
+        # Create a collection, find the samplesheet, and upload the reads
+        collection = self.res.collection.create(name='Test upload reads')
+        samplesheet = self.get_samplesheet()
+        collection.upload_reads(samplesheet)
+
+        # Check the uploaded data and created samples
+        self.assertEqual(len(collection.data), 2)
+        self.assertEqual(len(collection.samples), 2)
+        names = {sample.name for sample in collection.samples}
+        self.assertIn('single-reads', names)
+        self.assertIn('paired-reads', names)
+        sample1 = collection.samples.get(name='single-reads')
+        sample2 = collection.samples.get(name='paired-reads')
+        wait_process_complete(sample1.data[0], 1)
```

Comment:
Aren't tests waiting until processes are completed?

Reply:
There no need to wait for processes to be completed. We are also considering disabling processes from being run in e2e tests.

In ReSDK tests you should just check that inputs were saved correctly on the server. From there on you can rely on resolwe-bio tests that object will be processed correctly.

Reply:
This is not a desired behaviour. If server's queue is filled up this would mean that script can run for a long time (e.g. days), which is not acceptable for users.

Reply:
In practice, these upload scripts already sometimes take days to run, simply because of the size and quantity of the files. So, we tend to run them on a dedicated server using screen. 

The way I have this iteration written now, the demultiplex processes will all be triggered first, and the sample creation will happen in bulk afterwards. So, if the script gets interrupted, the data will be on the servers, but the samples will not be created appropriately. 

One solution to this is I could turn the demultiplexed sample creation into a separate function call that can be independently triggered on any collection with multiplexed data objects. So then the whole pipeline can be triggered at once, or one step at a time (upload, recreate samples, annotate).

Reply:
This is a really bad practice and once we will start using it there will be no way back.

The right way to handle this would be to make dss process which does everything. But this requires support for processes to fill sample (and data) descriptors. Until we support this, you should do the least ugly workaround, which in my opinion are independent functions.